client/src/pages/TwitterConnect.js

```diff
@@ -1,24 +1,41 @@
 import React, { useEffect } from 'react';
-import axiosInstance from '../api/buildClient';
+import { useDispatch, useSelector } from 'react-redux';
+import Loading from '../components/Loading';
+import { useQuery } from '../hooks/useQuery';
+import {
+  connectToTwitter,
+  getTwitterProfileInfoData,
+} from '../redux/actions/twitterActions';
+import { setUserMessages } from '../redux/actions/userActions';
 
 const TwitterConnect = () => {
+  const isConnected = useSelector((state) => state.twitterReducer.isConnected);
+  const dispatch = useDispatch();
+  const query = useQuery();
+
   useEffect(() => {
-    const initiateRequest = async () => {
-      const res = await axiosInstance.get('/api/auth/twitter/connect', {
-        headers: {
-          'Access-Control-Allow-Origin': '*',
-        },
-      });
-      if (res) {
-        window.location.assign(
-          `https://twitter.com/oauth/authorize?oauth_token=${res.data.requestToken}`
-        );
-      }
-    };
-    initiateRequest();
-  });
-  // TODO Create a Loading component
-  return <div>Loading</div>;
+    // if user already connected his twitter account, close tab
+    if (isConnected) {
+      window.close();
+    } else if (query.get('success') === 'true') {
```

Comment:
Do you use query to check if the link has true/false to see if it was a succesfull operation?
I suggest use redux/state

Reply:
Or even better dispatch getTwitterProfileInfoData when the connect twitter action is finished

Reply:
Yes, success=true/false is used to know if user authorized the app or did not (clicked cancel). 

In the backend, after user authorized or not the app, twitter sends a request to callback URL or not and I need to inform user if the connection was successful. Using redirect to the connect page and this success query I was able to make this whole process as smooth as possible.
I can't really use getTwitterProfileInfoData directly after connectToTwitter action as this is a request in the "middle" of the whole process, so to say. This will end up returning an error because the user hasn't yet approved or rejected the connection with Twitter.
The process goes like this: 
- user gets on mark.dev/twitter/connect (because there is no twitter info stored in redux, initiate connection)
- request to /api/auth/twitter/connect is made to API to get connection request tokens from Twitter.
- after we get the request tokens, change go to twitter/oauth/authorize and include request token in query.
- at this point the user seens the page with the 2 buttons to authorize or cancel connection. 
- depending on the result, twitter sends an error or auth tokens back to the callback URL provided in the backend while setting up oauth.
- if there are no errors, save access tokens in db and redirect user to mark.dev/twitter/connect with the success query set to true (or set to false if the user did not authorize)
- at this point we're on mark.dev/twitter/connect with success query set to true
- send request to get user twitter data and save in redux
- isConnected will get updated in redux, 
- TwitterConnect component re-renders and closes the window because user is connected. 


This was the best way i found to make the whole process smooth